crates/gazette/src/router.rs

```diff
@@ -0,0 +1,167 @@
+        };
+        let endpoint = tonic::transport::Endpoint::from_shared(endpoint.clone())
+            .map_err(|_err| Error::InvalidEndpoint(endpoint.clone()))?;
+
+        let client = (self.dialer)(endpoint).await?;
+        *state = Some((client.clone(), 1));
+
+        Ok(client)
+    }
+
+    fn pick(
+        &self,
+        route: Option<&broker::Route>,
+        primary: bool,
+    ) -> (Option<usize>, DialState<Client>) {
+        // Acquire non-async lock which *cannot* be held across an await point.
+        let mut states = self.states.lock().unwrap();
+        let index = pick(route, primary, &self.zone, &states);
+        //.unwrap_or((&default_id, &self.service_endpoint));
+
+        let default_id = MemberId::default();
+
+        let id = match index {
+            Some(index) => &route.unwrap().members[index],
+            None => &default_id,
+        };
+
+        let state = match states.get(id) {
+            Some(value) => value.clone(),
+            None => states.entry(id.clone()).or_default().clone(),
+        };
+
+        (index, state)
+    }
+
+    pub fn sweep(&self) {
+        let mut states = self.states.lock().unwrap();
+
+        states.retain(|id, state| {
+            // Retain entries which are currently connecting.
+            let Some(mut state) = state.try_lock() else {
+                return true;
+            };
+            // Drop entries which are not connected.
+            let Some((_client, uses)) = &mut *state else {
+                return false;
+            };
+            // Drop entries which have not been used since the last sweep.
+            if *uses == 0 {
+                tracing::debug!(?id, "dropping idle member connection");
+                return false;
+            }
+            *uses = 0; // Mark for next sweep.
+            true
+        });
+    }
+}
+
+fn pick<Client>(
+    route: Option<&broker::Route>,
+    primary: bool,
+    zone: &str,
+    states: &HashMap<MemberId, DialState<Client>>,
+) -> Option<usize> {
+    let default_route = broker::Route::default();
+    let route = route.unwrap_or(&default_route);
+
+    route
+        .members
+        .iter()
+        .zip(route.endpoints.iter())
+        .enumerate()
+        .max_by_key(|(index, (id, _endpoint))| {
+            let connected = if let Some(state) = states.get(id) {
+                if let Some(state) = state.try_lock() {
+                    if let Some(_conn) = state.as_ref() {
+                        true // Transport is ready.
+                    } else {
+                        false // Transport is not ready and no task is starting it.
+                    }
+                } else {
+                    true // Another task has started this transport.
+                }
+            } else {
+                false // Transport has not been started.
+            };
+
+            // Member selection criteria:
+            (
+                // If we want the primary, then prefer the primary.
```

Comment:
That's correct, but to keep routing infallible, rather than injecting an error here and then having to handle it, we just pick a broker that is available and let **it** return the error.

Oftentimes that error also includes a route update that the caller can use in its next attempt.